pkg/kv/kvserver/replica_application_result.go

```diff
@@ -225,6 +280,160 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
 	}
 }
 
+func (r *Replica) tryReproposeWithNewLeaseIndexV2(
+	ctx context.Context, origCmd *replicatedCmd,
+) error {
+	if origCmd.proposal.v2Reproposed {
+		// NB: we can also check v2SeenDuringApplication instead if we ever want to
+		// remove v2Reproposed. We should never hit this case: to hit it, we'd need
+		// multiple entries (i.e. copies of the same entry) of the same proposal
+		// coming up for application, each bound to a proposal. But since under V2
+		// the first entry that comes up takes the proposal out of the map (and it
+		// never re-enters the map after), the second copy cannot have reached this
+		// code path (since getting here means being associated to a ProposalData
+		// that was taken from the map).
+		return errors.AssertionFailedf("saw previously reproposed local command: %+v", origCmd.proposal)
+	}
+
+	// NB: `origCmd` remains "Local". It's just not going to signal anyone
+	// or release any latches.
+
+	origP := origCmd.proposal
+	if origP.v2Reproposed {
+		// In V2, a proposal is applied only once (see invariant in [1]). Since it
+		// has to be local to enter this branch, there is a bug - we should not see
+		// it as local twice in a row.
+		//
+		// TODO(tbg): it's likely that the `v2Reproposed` field isn't adding any
+		// value, since we already assert that v2SeenDuringApplication is false (see
+		// [1]).
+		//
+		// [1]: retrieveLocalProposalsV2
+		return errors.AssertionFailedf("proposal considered for reproposal twice")
+	}
+	origP.v2Reproposed = true
+
+	// We want to move a few items from origCmd to the new command, but only if we
+	// managed to propose the new command. For example, if we move the latches
+	// over too early but then fail to actually get the new proposal started, the
+	// old proposal will not release the latches. This would result in a lost
+	// latch.
+	var success bool
+
+	// Go through the original proposal field by field and decide what transfers
+	// to the new proposal (and how that affects the old proposal). The overall
+	// goal is that the old proposal remains a local proposal (switching it to
+	// non-local now invites logic bugs) but not bound to the caller.
+
+	// NB: quotaAlloc is always nil here, because we already
+	// released the quota unconditionally in retrieveLocalProposalsV2.
+	// So the below is a no-op.
+	//
+	// TODO(tbg): if we shifted the release of proposal quota to *after*
+	// successful application, we could move the quota over
+	// prematurely releasing it here.
+	newQuotaAlloc := origP.quotaAlloc
+	defer func() {
+		if success {
+			origP.quotaAlloc = nil
+		}
+	}()
+
+	newCommand := kvserverpb.RaftCommand{
```

Comment:
We should have a comment on `RaftCommand` pointing out that any changes to it must be reflected here. Alternatively (and better), use a method like we do during Raft application where we zero out the fields we've processed and then assert that the remaining struct is zero-values.

I see you mentioned a unit test below. That would also work.

Reply:
Going with the unit test for now. The semantics of what needs to be set when, etc, are tricky. What matters most is that we are forced to think about it when adding a new field.